File: DuplicateEncoder.js
// The goal of this exercise is to convert a string to a new string where each character in the new string is "(" if that character appears only once in the original string, or ")" if that character appears more than once in the original string. Ignore capitalization when determining if a character is a duplicate.

// Examples
// "din"      =>  "((("
// "recede"   =>  "()()()"
// "Success"  =>  ")())())"
// "(( @"     =>  "))((" 
// Notes
// Assertion messages may be unclear about what they display in some languages. If you read "...It Should encode XXX", the "XXX" is the expected result, not the input!


function duplicateEncode(word){
    let hash = {}
    let lower = word.toLowerCase()
    for(let letter of lower){
      hash[letter] = hash[letter] + 1 || 1
    }
      let answer = []
    for (let letter of lower){
      if(hash[letter] < 2){
        answer.push("(")
      } else {
        answer.push(")")
      }
    }
    return answer.join("")
      // ...
  }

  function duplicateEncode(word){
    return word
      .toLowerCase()
      .split('')
      .map( function (a, i, w) {
        return w.indexOf(a) == w.lastIndexOf(a) ? '(' : ')'
      })
      .join('');
  }

  /* 
  ok so my way was less efficient than the more clever way. so we are going to go over that instead. with the more cleaver way, 
  we are going to pretty much do the same thing. we are going to take the word and make it lowercase as specified by the problem.
  then we are going to split into an array and run a map function. a is the letter, i is the index, and w is the whole array.
  so we check the index of a in w with the last index. if its the same we return ( if its different we return ). simple as that.
  */

  

File: DuplicateEncoder.py

# The goal of this exercise is to convert a string to a new string where each character in the new string is "(" if that character appears only once in the original string, or ")" if that character appears more than once in the original string. Ignore capitalization when determining if a character is a duplicate.

# Examples
# "din"      =>  "((("
# "recede"   =>  "()()()"
# "Success"  =>  ")())())"
# "(( @"     =>  "))((" 
# Notes
# Assertion messages may be unclear about what they display in some languages. If you read "...It Should encode XXX", the "XXX" is the expected result, not the input!






def duplicate_encode(word):
    
    print(word)
 
    neword = word.lower()
    encode = ""
    
    for letter in neword:
        if word.count(letter) > 1:
            encode += ")"
        else:
            encode += "("
            
    return encode
    
    
    

def duplicate_encode(word):
    word = word.lower()  # Convert the word to lowercase for case-insensitive comparison
    
    encode = {}
    
    for letter in word:
        if letter in encode:
            encode[letter] += 1
        else:
            encode[letter] = 1
    
    encoded_string = ""
    
    for letter in word:
        if encode[letter] > 1:
            encoded_string += ")"
        else:
            encoded_string += "("
    
    return encoded_string



# for this we need to check the count of each letter. So we create a dict or we can use .count. .count is easier but both
# are relavant. so we create a for loop and itterate over our word checking if the neword.count(letter) is greater than 1.
# if it is, we add ), if not we add (.

#then return encode or encoded_string